### PR TITLE
Add ClusterError.

### LINKED ISSUE #244: Clumsy error if port cannot be open.
It is very common to not be able to open a port. For instance, root required for low values, or port already occupied.
We should have nice error message in that case. right now it shows:

```
Command failed: Artillery :: I/O error occurred: Permission denied (os error 13)
```

### PR DESCRIPTION
**Pull request by mosuka**
### Context / purpose
Closes #244 

### Description
Add ClusterError to display the cause of the error in an easy-to-understand manner.

### How was this PR tested?
Deliberately generated an error by running the following command and checked the message.

```
cargo run -- serve --host=0.0.0.0 --index-uri=file:///home/minoru/tmp/quickwit/indices/hdfslogs-idx --port=80
```

### Checklist
- [x] tested locally
- [x] included documentation

### CODE REVIEW

**Comment by mosuka** [comment]
```
% cargo run -- serve --host=0.0.0.0 --index-uri=file:///home/minoru/tmp/quickwit/indices/hdfslogs-idx --port=80
    Finished dev [unoptimized + debuginfo] target(s) in 0.13s
     Running `target/debug/quickwit serve --host=0.0.0.0 '--index-uri=file:///home/minoru/tmp/quickwit/indices/hdfslogs-idx' --port=80`
Jul 15 17:40:00.673  INFO quickwit_cluster::cluster: Read existing host key. host_key=57aa5a51-56ef-4d76-9880-16e749788a8c host_key_path="/tmp/quickwit-host-key-0.0.0.0-80"
Jul 15 17:40:00.673  INFO quickwit_cluster::cluster: Create new cluster. host_key=57aa5a51-56ef-4d76-9880-16e749788a8c listen_addr=0.0.0.0:81
Command failed: Failed to create cluster. Cause `Artillery :: I/O error occurred: Permission denied (os error 13)`
```

File: quickwit-cluster/src/error.rs [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+//
+//  You should have received a copy of the GNU Affero General Public License
+//  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+use thiserror::Error;
+
+/// Cluster error kinds.
+#[derive(Debug, Error)]
+pub enum ClusterError {
+    /// The target index already exists (Returned when creating a new index).
```
Comment (fulmicoton) [inline quickwit-cluster/src/error.rs:26]: This error message seems wrong.

File: quickwit-cluster/src/error.rs [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+    CreateClusterError {
+        /// Root cause.
+        cause: anyhow::Error,
+    },
+
+    /// Read host key error.
+    #[error("Failed to read host key. Cause: `{cause}`")]
+    ReadHostKeyError {
+        /// Root cause.
+        cause: anyhow::Error,
```
Comment (fulmicoton) [inline quickwit-cluster/src/error.rs:37]: add the  `#[source]` attribute to cause 

```
#[error("Failed to read host key. Cause: `{cause}`")]
    ReadHostKeyError {
        /// Root cause.
        #[source]
        cause: anyhow::Error,
      ...
```

File: quickwit-cluster/src/error.rs [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    },
+
+    /// Write host key error.
+    #[error("Failed to write host key. Cause: `{cause}`")]
+    WriteHostKeyError {
+        /// Root cause.
+        cause: anyhow::Error,
```
Comment (fulmicoton) [inline quickwit-cluster/src/error.rs:44]: add a source comment

File: quickwit-cluster/src/error.rs (hunk 1 above)
Comment (mosuka) [reply quickwit-cluster/src/error.rs]: Sorry. I forgot to correct it.

**Comment by mosuka** [comment]
@fulmicoton 
Updated the branch. Please take another look.

File: quickwit-cluster/src/cluster.rs [hunk 4]
```diff
@@ -88,15 +104,19 @@ impl Cluster {
...
         info!( host_key=?host_key, listen_addr=?listen_addr, "Create new cluster.");
         let config = ArtilleryClusterConfig {
             cluster_key: CLUSTER_ID.as_bytes().to_vec(),
             listen_addr,
             ..Default::default()
         };
         let (artillery_cluster, _) =
-            ArtilleryCluster::new_cluster(host_key, config).map_err(|err| anyhow::anyhow!(err))?;
+            ArtilleryCluster::new_cluster(host_key, config).map_err(|err| {
+                ClusterError::CreateClusterError {
```
Comment (fulmicoton) [inline quickwit-cluster/src/cluster.rs:116]: We can be more accurate it...

err is ArtilleryError.
Can you match case on it, and if it is an IOError, we could return an error accordingly. 

Something 

```
enum ClusterError {
    #[error("Failed to bind UDP port :{port} for the gossip membership algorithm. Cause: {cause}.")]
    UDPPortBindingError {
        port: u16,
        #[source]
        cause: io::Error
  }
  ...

}
```

Comment (mosuka) [reply quickwit-cluster/src/cluster.rs]: How about this?

```
% cargo run -- serve --host=0.0.0.0 --index-uri=file:///home/minoru/tmp/quickwit_data/indices/wikipedia-idx --port=80
    Finished dev [unoptimized + debuginfo] target(s) in 0.21s
     Running `target/debug/quickwit serve --host=0.0.0.0 '--index-uri=file:///home/minoru/tmp/quickwit_data/indices/wikipedia-idx' --port=80`
Jul 16 15:32:24.034  INFO quickwit_cluster::cluster: Read existing host key. host_key=57aa5a51-56ef-4d76-9880-16e749788a8c host_key_path="/tmp/quickwit-host-key-0.0.0.0-80"
Jul 16 15:32:24.034  INFO quickwit_cluster::cluster: Create new cluster. host_key=57aa5a51-56ef-4d76-9880-16e749788a8c listen_addr=0.0.0.0:81
Command failed: Failed to bind UDP port :81 for the gossip membership algorithm. Cause: Permission denied (os error 13)

Caused by:
    Permission denied (os error 13)
```

**Review by fulmicoton — APPROVED** [review]

**Pull request merged**